### PR TITLE
Implement JDBC connector for SQL [HZ-1141]

### PR DESCRIPTION
**Pull request by frant-hartm**
This PR adds a new connector for JDBC to Hazelcast SQL, which allows connecting to any database supporting the JDBC interface.

This is a prerequisite for HZ-1261.

## Changes to SQL engine
- `primaryKey` property added to `MappingField` - this is required for building UPDATE/DELETE queries
- SqlConnector
  - added `externalName` to resolveAndValidateFields
  - variant of `fullScanReader` that receives HazelcastTable
  - variant of `updateProcessor` that receives updates to fields as `RexNode` (`Map<String, RexNode> updates`)

## Status
This will be released as Beta feature in 5.2.

## Unresolved issues that are left to 5.3
- producing CompactGenericRecord directly from the connector to avoid
conversion from row to GenericRecord where GenericRecord is required (e.g. in HZ-1261)
- Pushdown of UPDATE/DELETE statements - currently the UPDATE/DELETE
statements get rewritten to SELECT + UPDATE/DELETE with WHERE clause on
PK. PK is required on the table for UPDATE/DELETE statements.
- `SqlConnector#createTable` is called for each statement, it's somewhat heavyweight, but I discovered this rather late
- haven't looked into schemas at all
- Security (needs support in security for external data stores)
- Timestamp support with sub-second precision

## Usage
Configure an external JdbcDataStore (feature added in #21716)
```
  external-data-store:
    mysql-database:
      class-name: com.hazelcast.datastore.JdbcDataStoreFactory
      properties:
        jdbcUrl: jdbc:mysql://dummy:3306
        username: xyz
        password: xyz
      shared: true
```
A table `people` in the MySQL database exists:
```
CREATE TABLE (id INT PRIMARY KEY, name VARCHAR(100))
```

We can create the following mapping in Hazelcast:

```
CREATE MAPPING people (
 id INT, 
 name VARCHAR 
) 
TYPE JDBC 
OPTIONS ( 
 'externalDataStoreRef'='mysql-database'
)
```
and we can query the table in Hazelcast:
```
SELECT * FROM people;
```

Or if we have the following mapping for an IMap:
```
CREATE MAPPING my_map ( 
  __key INT, 
  id INT, 
  name VARCHAR
) 
TYPE IMap 
OPTIONS (
    'keyFormat' = 'int',
    'valueFormat' = 'compact',
    'valueCompactTypeName' = 'person'
)
```
we can load all data from the table to the map:
```
INSERT INTO my_map SELECT id AS __key,id,name FROM people
```

Checklist:
- [x] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [x] Label `Add to Release Notes` or `Not Release Notes content` set
- [x] Request reviewers if possible
- [x] Send backports/forwardports if fix needs to be applied to past/future releases

This is a new feature in 5.2, no backports.

### CODE REVIEW

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/SelectProcessorSupplier.java [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+    @Override
+    public void init(@Nonnull Context context) throws Exception {
+        evalContext = ExpressionEvalContext.from(context);
+        query = buildQuery();
+    }
+
+    private String buildQuery() {
+        ExpressionTranslator translator = new ExpressionTranslator(evalContext, fields);
+        String select;
+        if (projection != null) {
```
Comment (TomaszGaweda) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/SelectProcessorSupplier.java:85]: What if...
```java
if (projection != null) {
select = projection.stream()
    .map(projectionExpression ->  translator.translate(projectionExpression))
    .collect(joining(", "));
} else {
...
}
```

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/SelectProcessorSupplier.java [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+                        return new JetSqlRow(evalContext.getSerializationService(), row);
+                    }
+            );
+            processors.add(processor);
+        }
+        return processors;
+    }
+
+    private Object convertValue(Object value) {
```
Comment (TomaszGaweda) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/SelectProcessorSupplier.java:139]: What about enums, do we support any automatic conversion?

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/IMapSqlConnector.java [hunk 3]
```diff
@@ -112,8 +112,8 @@ public boolean isStream() {
     public List<MappingField> resolveAndValidateFields(
             @Nonnull NodeEngine nodeEngine,
             @Nonnull Map<String, String> options,
-            @Nonnull List<MappingField> userFields
-    ) {
+            @Nonnull List<MappingField> userFields,
+            String externalName) {
```
Comment (TomaszGaweda) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/IMapSqlConnector.java:116]: `@Nonnull` for consistency maybe

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/DeleteJdbcSqlConnectorTest.java [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+package com.hazelcast.jet.sql.impl.connector.jdbc;
+
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import static com.hazelcast.jet.sql.impl.connector.jdbc.JdbcSqlConnector.OPTION_JDBC_URL;
+
+public class DeleteJdbcSqlConnectorTest extends JdbcSqlTestSupport {
```
Comment (TomaszGaweda) [inline hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/DeleteJdbcSqlConnectorTest.java:25]: Missing standard test annotations

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/InsertJdbcSqlConnectorTest.java [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+import com.hazelcast.sql.HazelcastSqlException;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Ignore;
+import org.junit.Test;
+
+import static com.hazelcast.jet.sql.impl.connector.jdbc.JdbcSqlConnector.OPTION_JDBC_URL;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+public class InsertJdbcSqlConnectorTest extends JdbcSqlTestSupport {
```
Comment (TomaszGaweda) [inline hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/InsertJdbcSqlConnectorTest.java:28]: ditto annotations

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlTestSupport.java [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+    }
+
+    public static class MySQLDatabaseProvider implements TestDatabaseProvider {
+
+        private String jdbcUrl;
+
+        @Override
+        public String createDatabase(String dbName) {
+            jdbcUrl = "jdbc:tc:mysql:8.0.29:///" + dbName + "?TC_DAEMON=true&sessionVariables=sql_mode=ANSI";
+//            jdbcUrl = "jdbc:tc:mysql:8.0.29:///" + dbName + "?TC_DAEMON=true,sessionVariables=sql_mode=''ANSI''";
```
Comment (TomaszGaweda) [inline hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlTestSupport.java:243]: In last work we had a convention, that if you need to temporarily comment out some code or do some hack, then you write `// TODO HZ-1141`. With this it's easy to find all things that supposed to be fixed before release. Maybe it's worth implementing it in Hazelcast as well?

**Review by TomaszGaweda — CHANGES_REQUESTED** [review]

**Comment by TomaszGaweda** [comment]
Mostly minor comments, but IMHO worth addressing :)

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/InsertProcessorSupplier.java [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+                    buildQuery(),
+                    ds,
+                    (PreparedStatement ps, JetSqlRow row) -> {
+                        for (int j = 0; j < row.getFieldCount(); j++) {
+                            // JDBC parameterIndex is 1-based, so j + 1
+                            ps.setObject(j + 1, row.get(j));
+                        }
+                    },
+                    false,
+                    BATCH_LIMIT
```
Comment (TomaszGaweda) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/InsertProcessorSupplier.java:78]: IMHO would be good to make it configurable by user

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/ExpressionTranslator.java [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+
+        if (!SUPPORTED_EXPRESSIONS.contains(expression.getClass())) {
+            throw new IllegalArgumentException("Unsupported predicate " + expression);
+        }
+    }
+
+    @SuppressWarnings("ReturnCount")
+    public String translate(Expression<?> expression0) {
+
+        // TODO this grew a bit more than expected, introduce e.g. visitor?
```
Comment (TomaszGaweda) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/ExpressionTranslator.java:67]: or introduce Java 19 switch pattern matching ;) 

(just joking)

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlConnector.java [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+    public List<String> getPrimaryKey(Table table0) {
+        JdbcTable table = (JdbcTable) table0;
+        return table.getPrimaryKeyList();
+    }
+
+    @Nonnull
+    @Override
+    public Vertex updateProcessor(@Nonnull DAG dag, @Nonnull Table table0,
+                                  @Nonnull Map<String, Expression<?>> updatesByFieldNames) {
+        JdbcTable table = (JdbcTable) table0;
```
[the remark this answers is not in the document]
Comment (frant-hartm) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlConnector.java:260]: I followed the convention from `IMapSqlConnector`

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlTestSupport.java (hunk 6 above)
Comment (frant-hartm) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlTestSupport.java:243]: This is a leftover. Will remove the commented line.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlConnector.java [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+
+        Map<String, Expression<?>> remappedUpdatesByFieldNames = new HashMap<>();
+        for (Entry<String, Expression<?>> entry : updatesByFieldNames.entrySet()) {
+            remappedUpdatesByFieldNames.put(table.getField(entry.getKey()).externalName(), entry.getValue());
+        }
+        List<String> pkFields = getPrimaryKey(table0)
```
[the remark this answers is not in the document]
Comment (frant-hartm) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlConnector.java:266]: 2d5d5da9

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcTable.java [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+            if (field.getName().equals(fieldName)) {
+                return (JdbcTableField) field;
+            }
+        }
+        throw new IllegalArgumentException("Unknown field with name " + fieldName);
+    }
+
+    public JdbcTableField getFieldByExternalName(String externalName) {
+        List<TableField> fields = getFields();
+        for (TableField field0 : fields) {
```
[the remark this answers is not in the document]
Comment (frant-hartm) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcTable.java:94]: 2d5d5da9

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/SelectProcessorSupplier.java (hunk 1 above)
Comment (frant-hartm) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/SelectProcessorSupplier.java:85]: :+1: nice

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/IMapSqlConnector.java (hunk 3 above)
Comment (frant-hartm) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/IMapSqlConnector.java]: 7a12041f

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/DeleteJdbcSqlConnectorTest.java (hunk 4 above)
Comment (frant-hartm) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/DeleteJdbcSqlConnectorTest.java:26]: They get inherited from the `SqlTestSupport` and `SimpleTestInClusterSupport`.

```
mvn test -PparallelTest
```
correctly runs them in the following phase:
```
[INFO] --- maven-surefire-plugin:2.22.2:test (default-test) @ hazelcast-sql ---
```
which uses multiple forks.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/InsertProcessorSupplier.java (hunk 7 above)
Comment (frant-hartm) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/InsertProcessorSupplier.java]: bd1812dd

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlTestSupport.java (hunk 6 above)
Comment (frant-hartm) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlTestSupport.java]: I have moved the `TestDatabaseProvider` and subclasses to `hazelcast` module test jar. I need it in extensions.
7f7e89c3

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/SelectProcessorSupplier.java (hunk 2 above)
Comment (frant-hartm) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/SelectProcessorSupplier.java:133]: This conversion is from objects returned by the JDBC driver to what we support in our SQL. I don't think we need any special handling for enums here.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/InsertProcessorSupplier.java [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+            );
+            processors.add(processor);
+        }
+        return processors;
+    }
+
+    private String buildQuery() {
+        StringBuilder sb = new StringBuilder();
+        sb.append("INSERT INTO ")
```
[the remark this answers is not in the document]
Comment (frant-hartm) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/InsertProcessorSupplier.java]: 1eea8b40bed

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/file/FileSqlConnector.java [hunk 13]
```diff
@@ -71,8 +71,8 @@ public boolean isStream() {
     public List<MappingField> resolveAndValidateFields(
             @Nonnull NodeEngine nodeEngine,
             @Nonnull Map<String, String> options,
-            @Nonnull List<MappingField> userFields
-    ) {
+            @Nonnull List<MappingField> userFields,
+            String externalName) {
```
[the remark this answers is not in the document]
Comment (frant-hartm) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/file/FileSqlConnector.java]: 7a12041f

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlConnector.java [hunk 14]
```diff
@@ -0,0 +1,322 @@
...
+                MappingField mappingField = new MappingField(f.name(), f.type());
+                mappingField.setPrimaryKey(dbField.primaryKey());
+                resolvedFields.add(mappingField);
+            }
+        }
+        return resolvedFields;
+    }
+
+    @Nonnull
+    private Map<String, DbField> readDbFields(@Nonnull Map<String, String> options, @Nonnull String externalName) {
```
[the remark this answers is not in the document]
Comment (frant-hartm) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlConnector.java]: bafe53eb

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlConnector.java [hunk 15]
```diff
@@ -0,0 +1,322 @@
...
+                ResultSet rs = statement.getResultSet();
+                ResultSetMetaData metaData = rs.getMetaData();
+                for (int i = 1; i <= metaData.getColumnCount(); i++) {
+                    String columnName = metaData.getColumnName(i);
+                    fields.put(columnName, new DbField(metaData.getColumnClassName(i),
+                            columnName, pkColumns.contains(columnName)));
+                }
+
+            } catch (SQLException e) {
+                throw new RuntimeException(e);
```
[the remark this answers is not in the document]
Comment (frant-hartm) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlConnector.java]: cdcb26af11e

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/infoschema/InfoSchemaConnector.java [hunk 16]
```diff
@@ -67,8 +67,8 @@ public boolean isStream() {
     public List<MappingField> resolveAndValidateFields(
             @Nonnull NodeEngine nodeEngine,
             @Nonnull Map<String, String> options,
-            @Nonnull List<MappingField> userFields
-    ) {
+            @Nonnull List<MappingField> userFields,
+            String externalName) {
         throw new UnsupportedOperationException();
```
[the remark this answers is not in the document]
Comment (TomaszGaweda) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/infoschema/InfoSchemaConnector.java:72]: Sure, just commented in spirit of "leave the place better than it was" ;)

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/ExpressionTranslator.java [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+public class ExpressionTranslator {
+
+    private static final Set<Class<?>> SUPPORTED_EXPRESSIONS;
+
+    static {
+        SUPPORTED_EXPRESSIONS = new HashSet<>();
```
[the remark this answers is not in the document]
Comment (frant-hartm) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/ExpressionTranslator.java]: The ExpressionTranslator class doesn't exist anymore.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/ExpressionTranslator.java [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+            return;
+        }
+
+        if (!SUPPORTED_EXPRESSIONS.contains(expression.getClass())) {
+            throw new IllegalArgumentException("Unsupported predicate " + expression);
+        }
+    }
+
+    @SuppressWarnings("ReturnCount")
+    public String translate(Expression<?> expression0) {
```
[the remark this answers is not in the document]
Comment (frant-hartm) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/ExpressionTranslator.java]: The ExpressionTranslator class doesn't exist anymore.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/ExpressionTranslator.java [hunk 19]
```diff
@@ -0,0 +1,130 @@
...
+import com.hazelcast.sql.impl.expression.predicate.ComparisonMode;
+import com.hazelcast.sql.impl.expression.predicate.ComparisonPredicate;
+import com.hazelcast.sql.impl.expression.predicate.OrPredicate;
+import com.hazelcast.sql.impl.expression.string.ConcatFunction;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+public class ExpressionTranslator {
```
[the remark this answers is not in the document]
Comment (frant-hartm) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/ExpressionTranslator.java]: The ExpressionTranslator class doesn't exist anymore.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/UpdateProcessorSupplier.java [hunk 20]
```diff
@@ -0,0 +1,124 @@
...
+
+    @Nonnull
+    @Override
+    public Collection<? extends Processor> get(int count) {
+        List<Processor> processors = new ArrayList<>(count);
+        for (int i = 0; i < count; i++) {
+            Processor processor = new WriteJdbcP<>(
+                    query,
+                    dataSource,
+                    (PreparedStatement ps, JetSqlRow row) -> {
```
Comment (TomaszGaweda) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/UpdateProcessorSupplier.java:83]: teeny-tiny suggestion: maybe worth moving the lambda to a method, so here you will have just `this::prepareStatement` and the lambda body will become method body. Same for other processors.
Nothing very important, just reading it is then easier when you have named methods (same for reading logs).

**Review by TomaszGaweda — APPROVED** [review]
Looks great! One tiny comment, but strictly style related, you can ignore if you don't agree.

**Review by Fly-Style — APPROVED** [review]
SQL part LGTM!

**Pull request merged**